File: problems/51/51.n-queens.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/*
 * @lc app=leetcode id=51 lang=java
 *
 * [51] N-Queens
 */

// @lc code=start
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        String[][] board = new String[n][n];
        Set<Integer> cols = new HashSet<>();
        Set<Integer> posDiags = new HashSet<>();
        Set<Integer> negDiags = new HashSet<>();

        backTrack(0, n, res, board, cols, posDiags, negDiags);
        return res;
    }

    private void backTrack(int row, int n, List<List<String>> res, String[][] board, Set<Integer> cols, Set<Integer> posDiags, Set<Integer> negDiags) {
        if (row == n) {
            res.add(new ArrayList<>(addBoard(board, n)));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || posDiags.contains(row + col) || negDiags.contains(row - col)) {
                board[row][col] = ".";
                continue;
            }

            cols.add(col);
            posDiags.add(row + col);
            negDiags.add(row - col);
            board[row][col] = "Q";

            backTrack(row + 1, n, res, board, cols, posDiags, negDiags);

            board[row][col] = ".";
            cols.remove(col);
            posDiags.remove(row + col);
            negDiags.remove(row - col);
        }
    }

    private List<String> addBoard(String[][] board, int n) {
        List<String> currBoard = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            String currRow = "";
            for (int j = 0; j < n; j++) {
                if (board[i][j] == null) {
                    board[i][j] = ".";
                }
                currRow += board[i][j];
            }
            currBoard.add(currRow);
        }

        return currBoard;
    }
}
// @lc code=end

File: problems/51/51.n-queens.py
#
# @lc app=leetcode id=51 lang=python3
#
# [51] N-Queens
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        # seperate rows, cols, diagonals
        # keep track of placed col, placed posDiag, placed negDiag
        # recursive through each row

        cols = set()
        posDiags = set() # r + c
        negDiags = set() # r - c
        res = []
        board = [["."] * n for _ in range(n)]
        
        def backtrack(curr_row):
            if curr_row == n:
                # solution reached
                curr_board = ["".join(row) for row in board]
                res.append(curr_board)
            
            for col in range(n):
                if col in cols or (curr_row + col) in posDiags or (curr_row - col) in negDiags:
                    continue

                cols.add(col)
                posDiags.add(curr_row + col)
                negDiags.add(curr_row - col)
                board[curr_row][col] = "Q"

                backtrack(curr_row + 1)

                cols.remove(col)
                posDiags.remove(curr_row + col)
                negDiags.remove(curr_row - col)
                board[curr_row][col] = "."

        backtrack(0)
        return res
        
# @lc code=end

